returnn/datasets/meta.py

```diff
@@ -1406,13 +1407,19 @@ def __init__(
           Now it could happen that ceildiv(data_len1 + data_len2, 6) < align_len1 + align_len2.
           This option would repeat intermediate ending frames such that data_len1 % 6 == 0,
           by setting it to {"data": 6}.
+        :param dict[str,(str,int)]|None pad_truncate_data_to_multiple_of: data_key -> (target_key, multiple).
+            Force data_len == target_len * multiple for specified data and target.
+            Similar to repeat_in_between_last_frame_up_to_multiple_of but also truncates the data.
+            Mainly for raw wav alignment, where data_len could be slightly longer than target_len * multiple
+            due to mismatch in feature extraction.
```

Comment:
"raw wav alignment" doesn't really say anything / is not relevant here.

What is the case you have specifically, i.e. what type of padding was done? This is what is relevant.

Why is `repeat_in_between_last_frame_up_to_multiple_of` not already enough?

Reply:
The main problem I want to tackle here is indeed relevant to "raw wav alignment"(but of course also to the padding scheme). The main problem with raw wav is that even with same padding, we could have $\left \lceil{T_{in}/F}\right \rceil \neq T_{out}$. With most common `stft` setting, we have `frame_size` of 25ms and `frame_shift` of 10 ms. This means a perfectly aligned 3-frame text corresponds to 45ms audio data, and you can see $\left \lceil{45/10}\right \rceil \neq 3$. In chunked training, we solve this by setting min_chunk_size to a magic number(which is not elegant, but we all do this, see [more explanation](https://omnifluent.atlassian.net/wiki/spaces/AS/pages/2636709904/Deep+dive+into+NN+training#Chunking)). But this gets wrong if such mismatch adds up with sequence concatenation. That's why I truncate such mismatches for all sub-sequences except the last one.

Reply:
> The main problem I want to tackle here is indeed relevant to "raw wav alignment"(but of course also to the padding scheme).

This doesn't make sense. It can not be relevant that you operate on raw wav. It only matters what kind of downsampling/padding scheme you have. Why don't you say what case you have actually? What code are you using? You can simply look that up.

> The main problem with raw wav is that even with same padding, we could have $\left \lceil{T_{in}/F}\right \rceil \neq T_{out}$

No. Padding=same implies that you have $\left \lceil{T_{\textrm{in}}/F}\right \rceil = T_{\textrm{out}}$, always. That follows directly from the definition of padding=same. That's actually what it means. If you don't have $\left \lceil{T_{\textrm{in}}/F}\right \rceil = T_{\textrm{out}}$, then it means, you don't have padding=same but you have some different padding scheme.

So I get back to my question: What padding scheme do you have?

> With most common `stft` setting, ...

I assume you have padding=valid then?

> In chunked training, we solve this by setting min_chunk_size to a magic number

Was this related to the discussion here? https://github.com/rwth-i6/returnn/issues/1564

But again, the main question: What type of padding do you use? Or if you don't know and cannot tell from your code, can you at least say what code you use?

Reply:
Hmm, I thought I knew which padding you were talking about but now I'm not sure. Do you refer to the padding/downsampling scheme in RETURNN? E.g. `pad_last_frame` in `MultiChannelStftLayer` or same padding in conv layer if conformer was used for downsampling?

Reply:
No, I mean whatever you did to get your original input len <-> output len relation. I don't care about what software you used. I just want to know your exact input len <-> output len relation, i.e. the type of padding scheme you have used there.

Reply:
But that's why raw wav is relevant here. In raw wav, the first frame is longer than the rest of frames due to mismatched window size and window shift step in feature extraction(see my example above). A padding can't fix this because it's a problem of too long data instead of too short data. That's why I propose to truncate in ConcatSeqsDataset in addition to the padding you implemented before.

Reply:
Sorry for the confusion, I had a bit discussion with Moritz and it turns out I didn't understand your point on padding. 
The main issue here is the padding logic in my RETURNN in-network feature extraction, as you described
> For padding=valid (i.e. no padding allowed for the window, the window must always cover real frames), you would have ⌈(Tin−(W−1))/F⌉=Tout, where W is the window/kernel size.

does not fit the padding logic in RASR alignment generation. So yes, it is possible to modify the alignment somewhere else to avoid the modification of RETURNN. Then maybe I should close this pull request?

Reply:
I still don't understand your padding scheme. What *exactly* is your input len <-> output len relation of the given data you have? There must be an *exact* formula to describe that relation. If you don't know the formula, please find out. It would also help when you say what code in your case determines this relation. E.g. what is your feature extraction? I didn't really understand whether you use RETURNN, RASR, or sth else to generate this data, and in any case, what exactly.

Whatever the formula is, from the formula, you can infer your padding scheme. Nothing else is relevant. It's still not relevant that you operate on raw wav.

Reply:
Yes, I understand your point that operating on raw WAV is not relevant now, only the alignment logic applied there is relevant. Sorry again for the misunderstanding before. The exact formula in my setup(slightly different from most common setup): 
$T_{out}=\left \lfloor{(T_{in}-(W-F))/F} \right \rfloor$
, which is different from the alignment(or padding) scheme of RASR-generated alignment.

Reply:
Really floordiv, not ceildiv? Can you give a reference to the code which leads to such formula?

Reply:
Assuming the formula is correct: You can translate this into the standard scheme of the formula:
$T_{out} = \lfloor (T_{in}-(W-F))/F \rfloor = \lceil (T_{in} - (W - F) - (F - 1)) / F \rceil = \lceil (T_{in} - (W - 1)) / F \rceil$
I.e. it means you have exactly the padding=valid case. That's what I wanted to know.

Reply:
Yes, padding=valid is the case I have. Then what would be your suggestion in this case? Set up the alignment so that padding=valid also holds there instead of modifying RETURNN? Thanks!

Reply:
No. I was simply saying, you should clarify the documentation: you should make clear, for what cases the user would want to use `pad_truncate_data_to_multiple_of`. Why is `repeat_in_between_last_frame_up_to_multiple_of` not enough?

See my last comment in the main thread. The answer is basically: `repeat_in_between_last_frame_up_to_multiple_of` only works for padding=same. But your option works for any padding scheme, e.g. for padding=valid.

Additionally, as you misunderstood for a while this concept of padding scheme (no matter what type of input/output, no matter the software, the input len <-> output len relation can always be expressed as a formula, and that always corresponds to a specific padding scheme), it probably would be helpful to add another sentence or so explaining this.